### ISSUE TITLE
Puppeteer leaks a lot of registered CDP event handlers due to missing domain destructors

### ISSUE THREAD

**Issue by whimboo** [puppeteer/puppeteer#5043]
**Tell us about your environment:**

* Puppeteer version: 1.20.0-post (most recent checkout from master)
* Platform / OS version: MacOS
* Node.js version: v12.1.0

**What steps will reproduce the problem?**
I have seen this problem when running the puppeteer unit tests against Firefox with leak checks turned on. While our internal browser chrome tests which check our CDP implementation works fine, all the puppeteer unit tests are failing. As such Puppeteer itself is leaking lots o memory.

_Please include code that reproduces the issue._
No specific code is necessary here.  It's all under https://github.com/GoogleChrome/puppeteer/blob/master/test/. Just run the tests.

Specifically the [puppeteer test runner](https://github.com/GoogleChrome/puppeteer/blob/master/test/puppeteer.spec.js#L107-L117) demonstrates that very well. As it can be seen for each and every test a new instance of the Page domain is getting created, and set to `null` afterward. Basically that should return some kind of destructor code to clean-up various domains and registered events. But there is none. Here an example:

1. The page domain [creates a new FrameManager](https://github.com/GoogleChrome/puppeteer/blob/master/lib/Page.js#L70) instance.
2. The FrameManager [creates a new NetworkManager](https://github.com/GoogleChrome/puppeteer/blob/master/lib/FrameManager.js#L38) instance, and [initializes](https://github.com/GoogleChrome/puppeteer/blob/master/lib/FrameManager.js#L67) it.
3. Initialization of the NetworkManager calls the [Network.enable](https://github.com/GoogleChrome/puppeteer/blob/master/lib/NetworkManager.js#L59) method of the CDP `Network` domain.
4. [Network.enable](https://chromedevtools.github.io/devtools-protocol/tot/Network#method-enable) is specified as: 

> Enables network tracking, network events will now be delivered to the client.

5. Given that `Network.disable` is never called, all the registered events for intercepting the network communication inside of the CDP domain are leaking.

@mathiasbynens would you mind to have a look at this? Thanks

**Comment by whimboo** [comment]
@OrKoN I wonder if there is something that could be done for this issue?

**Comment by OrKoN** [comment]
@whimboo I think this might be very relevant and hopefully solved with the ongoing refactoring. Although I am not sure: is it specific to Firefox? I think in Chrome when a session detaches the agents get destroyed on the backend and the session listeners get cleaned up to.

**Comment by OrKoN** [comment]
@whimboo how did you enable the leak checks to find this?

**Comment by whimboo** [comment]
I tried to reproduce but wasn't successful yet. Basically someone would need a debug build of Firefox and set the `XPCOM_MEM_BLOAT_LOG=1` environment variable when running tests with Firefox. But all the output I can see is not related to our CDP implementation.

As such I wonder if with the refactoring of the repository from just JS code to TypeScript the problematic parts have been fixed.

Shall we close the issue?
